From 64923d075132dfb81d75436bb08711f16beebe26 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Tue, 8 Aug 2017 14:00:26 +0800
Subject: [PATCH 2348/5242] MLK-16158-3 video: fbdev: mipi_dsi_northwest: add
 'vmode_index' to help phy pll config

commit  1f2fed5d241476b204e490b4dd92a239cdbac005 from
https://source.codeaurora.org/external/imx/linux-imx.git

Add a field 'vmode_index' to 'mipi_dsi_info' structure to
save the display video mode derived from dts to help config
the PHY PLL as desired.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/mipi_dsi.h           |    1 +
 drivers/video/fbdev/mxc/mipi_dsi_northwest.c |   22 ++++++++++++++++++----
 2 files changed, 19 insertions(+), 4 deletions(-)

diff --git a/drivers/video/fbdev/mxc/mipi_dsi.h b/drivers/video/fbdev/mxc/mipi_dsi.h
index 6b957b9..b8d00b3 100644
--- a/drivers/video/fbdev/mxc/mipi_dsi.h
+++ b/drivers/video/fbdev/mxc/mipi_dsi.h
@@ -100,6 +100,7 @@ struct mipi_dsi_info {
 	struct clk			*esc_clk;
 #endif
 	struct mxc_dispdrv_handle	*disp_mipi;
+	int				vmode_index;
 	struct  fb_videomode		*mode;
 	struct regulator		*disp_power_on;
 	struct  mipi_lcd_config		*lcd_config;
diff --git a/drivers/video/fbdev/mxc/mipi_dsi_northwest.c b/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
index 1aad721..278e55e 100644
--- a/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
+++ b/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
@@ -332,10 +332,23 @@ static int mipi_dsi_dphy_init(struct mipi_dsi_info *mipi_dsi)
 		div.co = 0x0;  /* 1  */
 	} else {
 #ifdef CONFIG_FB_IMX64
-		/* pll vco = 27 * 33 / (1 * 2) = 445.5MHz */
-		div.cn = 0x1f; /* 1 */
-		div.cm = 0xc1; /* 33 */
-		div.co = 0x1;  /* 2 */
+		switch (mipi_dsi->vmode_index) {
+		case 34:	/* 1920x1080@30Hz */
+			/* pll vco = 27 * 33 / (1 * 2) = 445.5MHz */
+			div.cn = 0x1f; /* 1 */
+			div.cm = 0xc1; /* 33 */
+			div.co = 0x1;  /* 2 */
+			break;
+		case 16:	/* 1920x1080@60Hz */
+			/* pll vco = 27 * 33 / (1 * 1) = 891MHz */
+			div.cn = 0x1f; /* 1 */
+			div.cm = 0xc1; /* 33 */
+			div.co = 0x0;  /* 1 */
+			break;
+		default:
+			/* TODO: not support yet */
+			return -EINVAL;
+		}
 #else
 		/* pll vco = 24 * 63 / (5 * 1) = 302.4MHz */
 		div.cn = 0x1C; /* 5  */
@@ -1088,6 +1101,7 @@ static int mipi_dsi_probe(struct platform_device *pdev)
 		ret = of_property_read_u32(remote, "video-mode", &vmode_index);
 		if ((ret < 0) || (vmode_index >= ARRAY_SIZE(mxc_cea_mode)))
 			return -EINVAL;
+		mipi_dsi->vmode_index = vmode_index;
 
 		mipi_dsi->mode = devm_kzalloc(&pdev->dev,
 					      sizeof(struct fb_videomode),
-- 
1.7.9.5

